### ISSUE TITLE
WCF Server Connection Resets

### ISSUE THREAD

**Issue by cbenard** [dotnet/wcf#3771]
Could anyone clarify exactly what are the functions of [`MaxPendingAccepts`](https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.channels.connectionorientedtransportbindingelement.maxpendingaccepts?view=netframework-4.8) and [`MaxPendingConnections`](https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.channels.connectionorientedtransportbindingelement.maxpendingconnections?view=netframework-4.8) on the server side? I've read the docs, and I think this could possibly be related to a problem we're experiencing.

I think the `MaxPendingAccepts` docs may have a problem?

> “This property limits the number of channels that the server can have waiting on a listener. When `MaxPendingAccepts` is too low, there will be a small interval of time in which all of the waiting channels have started servicing connections, but no new channels have begun listening. A connection can arrive during this interval and will fail because nothing is waiting for it on the server. This property can be configured by setting the `MaxPendingConnections` property to a larger number.”

Does it really mean to change `MaxPendingAccepts` by changing `MaxPendingConnections`? I'm not clear on the difference between the two properties. We aren't currently setting `MaxPendingConnections` which [appears to default](https://docs.microsoft.com/en-us/dotnet/framework/wcf/wcf-simplification-features) to 12 * cores (8 in our case).

We're using net.tcp binding on a high-volume service. Our users have increased dramatically over the past few years and it's an NLB cluster of 4 servers, but sometimes we get bursts of traffic where (according to perfmon) calls per sec goes up to like 200+.

We are experiencing problems during this time where network traffic monitoring shows that we get a SYN and then within 0.2 ms we send back an ACK, RST closing the connection.

WCF tracing shows no errors during this time, so I didn't think it was throttling. I had formed a hypothesis that it was [SynAttackProtect](https://social.technet.microsoft.com/Forums/Lync/en-US/c53806fa-3c62-4b2b-a85a-fb1047a231bd/sync-attack-protection?forum=winserversecurity), but so far, I can't find anything positively or negatively confirming that in the netsh trace + Network Monitor.

Perfmon shows us well under our max instances, calls, and sessions.

Right now, our server config is this:
```MaxConcurrentCalls = 1500;
MaxConcurrentSessions = 1500;
MaxConcurrentInstances = 1500;
MaxConnections = 1600;
ListenBacklog = 800;
MaxPendingSessions = 300;
MaxPendingAccepts = 20;

// MaxPendingConnections = not set, as mentioned above,
// and docs seem to indicate this is 12 * cores (8).
```

My new hypothesis is that the reason we could be sending an RST in response to the SYN is because maybe there is no actual TCP listener? Would MaxPendingAccepts being less than the concurrent calls cause this behavior?

Should MaxPendingAccepts be cranked up to 300 like the pending sessions? Should MaxPendingConnections be set? Why do the docs say to set MaxPendingConnections in order to affect MaxPendingAccepts?

I am just trying to validate if refusing connections could be caused by the MaxPendingAccepts being 20.

**Comment by mconnew** [comment]
This repo is for the WCF client on .NET Core and doesn't include an implementation of server. I'll give a quick answer though. The documentation is correct. A pending connection is a connection where the socket has been connected and the connection handshake is underway. The connection handshake includes communicating what endpoint you are connecting to, any security upgrades (e.g. wrapping connection with NegotiateStream if using Windows auth) etc. The number of pending accepts will be the lower of MaxPendingAccepts and (MaxPendingConnections - current pending connections). The reason is if you have MaxPendingConnections=100 and currently have 100 connections being established, if you also then had a pending accept, what would you do with the socket once it's connected? You can't begin the connection handshake as you are at your limit. In order to have a pending accept, you must have capacity available in your pending connections.  
Having a ListenBacklog set to 800 should prevent any connection resets. The fact that you are makes me suspect you have TcpPortSharing enabled where SmSvcHost.exe handles accepting new connections. If this is the case, either stop using TcpPortSharing (not an option if IIS hosted), or modify your SmSvcHost.exe.config file.  
Here's a [good blog post](https://blogs.msdn.microsoft.com/asiatech/2012/07/16/modifying-smsvchost-exe-config-for-wcf-some-common-mistakes/) which should provide you with more than enough info if this is the case.  
As for what values to use, cranking up MaxPendingAccepts won't do anything for you if there's no pending connection slots available. MaxPendingAccepts should only need to be set to a low multiple of number of cores. . Theoretically this only needs to be the same as number of cores but software isn't perfectly written so the default is 2X the number of cores. You really shouldn't need to modify this.  
MaxPendingConnections needs to be configured for your connection rate. If you want to be mathematically about it, you take the connections per second rate that you want to be able to handle and multiply it by the average time it takes for a client to complete the handshake in seconds (so 10ms would be 0.010 seconds) and that tells you the concurrency you need. You don't need to set this high enough to handle your burst rate, just the rate you want to handle. If you set it too high, you actually slow everyone down and you increase CPU contention and it can end up taking more resources to do. CPU contention means switching between connections being established before completing the handshake and coming back to it later, and that switching is expensive.  
ListenBacklog is where you tune things for your burst rate. Basically the SYN packet arrives and the OS parks the connection request in a queue in the kernel. When the next Accept call happens, the OS picks the top one off the top of the queue. Depending on how long and big your bursts are, it can be fine to have a client wait a second or two before the connection is picked up. It's when the ListenBacklog is full that a RST is sent. This is why I suspect you are going through SmSvcHost, because with a ListenBacklog of 800, you shouldn't even be sending a RST. The default ListenBacklog is 12 * cores, so 96 in your case. Basically your ListenBacklog needs to be the different between your burst rate and your connection completion rate.

**Comment by cbenard** [comment]
@mconnew Thank you for your thorough reply. We aren't IIS hosted and TcpPortSharing is not enabled. `netstat -abn` bears this out too (proc name/port changed to protect the innocent):

```
  TCP    0.0.0.0:4321           0.0.0.0:0              LISTENING
 [OurServer.exe]
```

We optimized a factor that caused a lot of clients to connect at the same time, decreasing our burst rate, so we've mostly mitigated the symptom for now, but we could reproduce the problem at will.

If we continue to get RSTs with `ListenBacklog` above our burst rate, should we use a Microsoft Support case to get engineering support with WCF server in this case?
